node/src/key_events.rs

```diff
@@ -1,395 +1,494 @@
+            )
+        })
+        .await
+        .expect("Should not fail since closure does not panic")
+        .clone()
+}
+
+/// Resolves as soon as the contract has moved past the key event ID.
+async fn key_event_id_expiration(
+    mut key_event_receiver: watch::Receiver<ContractKeyEventInstance>,
+    key_event_id: KeyEventId,
+) {
+    key_event_receiver
+        .wait_for(|contract_event| {
+            matches!(
+                contract_event.compare_to_expected_key_event_id(&key_event_id),
+                KeyEventIdComparisonResult::ContractIsPastExpected
+            )
+        })
+        .await
+        .expect("Should not fail since closure does not panic");
 }
 
-fn convert(public_key: &near_sdk::PublicKey) -> anyhow::Result<AffinePoint> {
+fn sdk_public_key_to_affine_point(public_key: &near_sdk::PublicKey) -> anyhow::Result<AffinePoint> {
     let public_key = near_crypto::PublicKey::from_str(&String::from(public_key));
     public_key_to_affine_point(public_key.unwrap())
 }
 
-pub struct ResharingKeys {
-    pub previous_keyset: Keyset,
-    pub existing_keyshares: Vec<Keyshare>,
-    pub keyshare_storage: KeyshareStorage,
+/// The leader waits for at most this amount of time for the start transaction to materialize,
+/// before retrying.
+const MAX_LATENCY_BEFORE_EXPECTING_TRANSACTION_TO_FINALIZE: Duration = Duration::from_secs(20);
+
+/// The leader logic for an entire key generation (initializing) state.
+/// Handles multiple domains and attempts. It does not return, except in case of catastrophic
+/// failure (node shutting down). The coordinator is expected to interrupt this when the
+/// contract state transitions out of the key generation state.
+pub async fn keygen_leader(
+    client: Arc<MeshNetworkClient>,
+    keyshare_storage: Arc<KeyshareStorage>,
+    mut key_event_receiver: watch::Receiver<ContractKeyEventInstance>,
+    chain_txn_sender: mpsc::Sender<ChainSendTransactionRequest>,
+    threshold: usize,
+) -> anyhow::Result<()> {
+    loop {
+        // Wait for all participants to be connected. Otherwise, computations are most likely going
+        // to fail so don't waste the effort.
+        client.leader_wait_for_all_connected().await?;
+
+        // Wait for the contract to have no active key event instance.
+        let key_event_id = key_event_receiver
+            .wait_for(|contract_event| !contract_event.started)
+            .await?
+            .id;
+        // Send txn to start the keygen instance. This may or may not end up in the chain; we'll
+        // wait for it. If it doesn't happen after some time, we try again.
+        chain_txn_sender
+            .send(ChainSendTransactionRequest::StartKeygen(
+                ChainStartKeygenArgs { key_event_id },
+            ))
+            .await?;
+
+        match timeout(
+            MAX_LATENCY_BEFORE_EXPECTING_TRANSACTION_TO_FINALIZE,
+            key_event_receiver.wait_for(|contract_event| contract_event.id == key_event_id),
```

Comment:
If `contract_event.id` changes because the existing event timed out, will we need to wait for `MAX_LATENCY_BEFORE_EXPECTING_TRANSACTION_TO_FINALIZE` to pass before exiting this function?

Reply:
Ah I see, good point. I'll make this wait for the event to start instead, and after that, check if the ID is the same.